app/controllers/api/databases.php

```diff
@@ -2465,16 +2629,99 @@ function updateAttribute(
         $data['$collection'] = $collection->getId(); // Adding this param to make API easier for developers
         $data['$id'] = $documentId == 'unique()' ? ID::unique() : $documentId;
         $data['$permissions'] = $permissions;
+        $document = new Document($data);
 
-        try {
-            $document = $dbForProject->createDocument('database_' . $database->getInternalId() . '_collection_' . $collection->getInternalId(), new Document($data));
-            $document->setAttribute('$collectionId', $collectionId);
-            $document->setAttribute('$databaseId', $databaseId);
-        } catch (StructureException $exception) {
-            throw new Exception(Exception::DOCUMENT_INVALID_STRUCTURE, $exception->getMessage());
-        } catch (DuplicateException $exception) {
-            throw new Exception(Exception::DOCUMENT_ALREADY_EXISTS);
-        }
+        $checkPermissions = function (Document $collection, Document $document, string $permission) use (&$checkPermissions, $dbForProject, $database) {
+            $documentSecurity = $collection->getAttribute('documentSecurity', false);
+            $validator = new Authorization($permission);
+
+            $valid = $validator->isValid($collection->getPermissionsByType($permission));
+            if (($permission === Database::PERMISSION_UPDATE && !$documentSecurity) || !$valid) {
+                throw new Exception(Exception::USER_UNAUTHORIZED);
+            }
+
+            if ($permission === Database::PERMISSION_UPDATE) {
+                $valid = $valid || $validator->isValid($document->getUpdate());
+                if ($documentSecurity && !$valid) {
+                    throw new Exception(Exception::USER_UNAUTHORIZED);
+                }
+            }
+
+            $relationships = \array_filter(
+                $collection->getAttribute('attributes', []),
+                fn ($attribute) => $attribute->getAttribute('type') === Database::VAR_RELATIONSHIP
+            );
+
+            foreach ($relationships as $relationship) {
+                $related = $document->getAttribute($relationship->getAttribute('key'));
+
+                if (empty($related)) {
+                    continue;
+                }
+
+                $relatedCollectionId = $relationship->getAttribute('relatedCollection');
+                $relatedCollection = Authorization::skip(
+                    fn() => $dbForProject->getDocument('database_' . $database->getInternalId(), $relatedCollectionId)
+                );
+
+                foreach ($related as $document) {
```

Comment:
`$related` can be a single Document or a string, right? I don't think the permission check would run if a Document or string were passed.

Reply:
Looks like I get a warning like:

```
Warning: foreach() argument must be of type array|object, string given in /usr/src/code/app/controllers/api/databases.php on line 2667
```